src/main/resources/org/jenkinsci/plugins/workflow/job/console/NewNodeConsoleNote/script.js

```diff
@@ -1,41 +1,62 @@
-    var label = e.getAttribute('label')
+var enclosings = new Map() // id → enclosingId
+var labels = new Map() // id → label
+
+/**
+ * Fix label attributes
+ */
+function fixLabels(pipelineElement) {
+    var label = pipelineElement.getAttribute('label')
     if (label != null) {
-        var html = e.innerHTML
+        var html = pipelineElement.innerHTML
         var suffix = ' (' + label.escapeHTML() + ')';
         if (html.indexOf(suffix)<0) {
-            e.innerHTML = e.innerHTML.replace(/.+/, '$&' + suffix) // insert before EOL
+            pipelineElement.innerHTML = pipelineElement.innerHTML.replace(/.+/, '$&' + suffix) // insert before EOL
         }
-    }
-    var nodeId = e.getAttribute('nodeId')
-    var startId = e.getAttribute('startId')
-    if (startId == null || startId == nodeId) {
-        e.innerHTML = e.innerHTML.replace(/.+/, '$&<span class="pipeline-show-hide"> (<a href="#" onclick="showHidePipelineSection(this); return false">hide</a>)</span>')
-        // TODO automatically hide second and subsequent branches: namely, in case a node has the same parent as an earlier one
-    }
-    // The CSS rule for branch names only needs to be added once per node, so we
-    // check in case we are viewing the truncated log and have already processed
-    // a duplicate synthetic span element for this node.
-    var maybeDupeNodes = $$('[nodeid=\"'+nodeId+'\"].pipeline-new-node');
-    for (var i = 0; i < maybeDupeNodes.length; i++) {
-        var node = maybeDupeNodes[i];
-        if (node !== e && node.processedNewNodeConsoleNote) {
-            return;
-        }
-    }
-    var nodes = $$('.pipeline-new-node')
+     }
+}
+
+/**
+ * Generate hide/show links for all of the pipeline elements
+ */
+function addLinks (pipelineElement) {
+       var nodeId = pipelineElement.getAttribute('nodeId')
+       var startId = pipelineElement.getAttribute('startId')
+       if (startId == null || startId == nodeId) {
+            pipelineElement.innerHTML = pipelineElement.innerHTML.replace(/.+/, '$&<span class="pipeline-show-hide"> (<a href="#" onclick="showHidePipelineSection(this); return false">hide</a>)</span>')
+       }
+}
+
+/**
+ * Generate enclosings for a element
+ */
+function generateEnclosing (node) {
+    var oid = node.getAttribute('nodeId')
+    enclosings.set(oid, node.getAttribute('enclosingId'))
+    labels.set(oid, node.getAttribute('label'))
+}
+
+/**
+ * Generate enclosings and labels for all of the pipeline elements
+ */
+function generateEnclosings () {
     var enclosings = new Map() // id → enclosingId
     var labels = new Map() // id → label
+    var nodes = $$(".pipeline-new-node")
+
     for (var i = 0; i < nodes.length; i++) {
         var node = nodes[i]
         var oid = node.getAttribute('nodeId')
+
         enclosings.set(oid, node.getAttribute('enclosingId'))
         labels.set(oid, node.getAttribute('label'))
+        appendBranchNames(oid, enclosings, labels)
```

Comment:
It looks like this is the only place where `appendBranchNames` is called, which seems wrong, since this method is only called for the `parallel` step itself. What if we start executing the first branch in a `parallel` step, and then that branch stops and we start executing the other branch? The first branch might get labelled correctly, but I think that the second branch wouldn't be labeled (unless later on in the Pipeline there was another `parallel` step).